internal/provider/providers/hetzner/provider.go

```diff
@@ -0,0 +1,326 @@
+			ID    string `json:"id"`
+			Value netip.Addr `json:"value"`
+		} `json:"records"`
+	}{}
+	err = decoder.Decode(&listRecordsResponse)
+	if err != nil {
+		return "", false, fmt.Errorf("json decoding response body: %w", err)
+	}
+
+	switch {
+	case len(listRecordsResponse.Records) == 0:
+		return "", false, fmt.Errorf("%w", errors.ErrReceivedNoResult)
+	case len(listRecordsResponse.Records) > 1:
+		return "", false, fmt.Errorf("%w: %d instead of 1",
+			errors.ErrResultsCountReceived, len(listRecordsResponse.Records))
+	}
+	identifier = listRecordsResponse.Records[0].ID
+	upToDate = listRecordsResponse.Records[0].Value == newIP.String()
+	return identifier, upToDate, nil
+}
+
+func (p *Provider) createRecord(ctx context.Context, client *http.Client, ip netip.Addr) (recordID string, err error) {
+	recordType := constants.A
+	if ip.Is6() {
+		recordType = constants.AAAA
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.hetzner.com",
+		Path:   fmt.Sprintf("/api/v1/records"),
+	}
+
+	requestData := struct {
+		Type           string `json:"type"` 
+		Name           string `json:"name"` 
+		Value          string `json:"value"`
+		ZoneIdentifier string `json:"zone_id"`
+		TTL            uint   `json:"ttl"`
+	}{
+		Type:           recordType,
+		Name:           p.host,
+		Value:          ip.String(),
+		ZoneIdentifier: p.zoneIdentifier,
+		TTL:            p.ttl,
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	encoder := json.NewEncoder(buffer)
+	err = encoder.Encode(requestData)
+	if err != nil {
+		return "", fmt.Errorf("JSON encoding request data: %w", err)
+	}
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), buffer)
+	if err != nil {
+		return "", fmt.Errorf("creating http request: %w", err)
+	}
+
+	p.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode > http.StatusUnsupportedMediaType {
+		return "", fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
+
+	decoder := json.NewDecoder(response.Body)
+	var parsedJSON struct {
+		Record struct {
+			ID string `json:"id"`
+		} `json:"record"`
+	}
+	err = decoder.Decode(&parsedJSON)
+	if err != nil {
+		return "", fmt.Errorf("json decoding response body: %w", err)
+	}
+
+	if !parsedJSON.Record.ID {
+		return "", fmt.Errorf("%w: %s", errors.ErrUnsuccessful, "error: empty response without id")
+	}
+
+	return parsedJSON.Record.ID, nil
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+	recordType := constants.A
+	if ip.Is6() {
+		recordType = constants.AAAA
+	}
+
+	identifier, upToDate, err := p.getRecordID(ctx, client, ip)
+
+	switch {
+	case stderrors.Is(err, errors.ErrReceivedNoResult):
+		identifier, err = p.createRecord(ctx, client, ip)
+		if err != nil {
+			return netip.Addr{}, fmt.Errorf("creating record: %w", err)
+		}
```

Comment:
I think we can return if we create the record right?

```suggestion
		identifier, err = p.createRecord(ctx, client, ip)
		if err != nil {
			return netip.Addr{}, fmt.Errorf("creating record: %w", err)
		}
		return ip, nil
```